File: src/analysis/html/htmlDomParser.ts
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import { resolveAnalyzerScriptPath } from '../utils/analysisUtils';
import { executeCommand } from '../../pythonEnv/utils/processUtils';
import { getPythonExecutable, getVenvPath } from '../../pythonEnv/utils/pathUtils';

/**
 * Interface for DOM element representation
 */
export interface DOMElement {
  tagName: string;
  attributes: Record<string, string>;
  children: DOMElement[];
  textContent?: string;
  depth: number;
  id?: string;
  classes: string[];
}

/**
 * Interface for DOM analysis result
 */
export interface DOMAnalysisResult {
  fileName: string;
  filePath: string;
  totalElements: number;
  maxDepth: number;
  htmlContent: string;
  domTree: DOMElement;
  elementCounts: Record<string, number>;
  timestamp: string;
}

/**
 * Calls the Python script to parse the HTML file and returns the analysis result.
 * @param filePath Path to the HTML file
 * @returns The result from the Python script
 */
async function callPythonDOMParser(filePath: string): Promise<any> {
  // Create an output channel for debugging
  const outputChannel = vscode.window.createOutputChannel('CodeXR DOM Analysis');
  
  const pythonScriptPath = resolveAnalyzerScriptPath('html_dom_parser.py', outputChannel);
  outputChannel.appendLine(`Resolved script path: ${pythonScriptPath}`);
  
  // Verify the script exists
  if (!fs.existsSync(pythonScriptPath)) {
    outputChannel.appendLine(`❌ Script does not exist at: ${pythonScriptPath}`);
    throw new Error(`Python script not found at: ${pythonScriptPath}`);
  }
  
  // First try with system Python if virtual environment is not available
  let pythonExecutable: string;
  let usingVenv = false;
  
  const venvPath = getVenvPath();
  outputChannel.appendLine(`Virtual environment path: ${venvPath}`);
  
  if (venvPath && fs.existsSync(venvPath)) {
    const venvPython = getPythonExecutable(venvPath);
    if (fs.existsSync(venvPython)) {
      pythonExecutable = venvPython;
      usingVenv = true;
    } else {
      // Fallback to system Python
      pythonExecutable = process.platform === 'win32' ? 'python.exe' : 'python3';
    }
  } else {
    // Use system Python
    pythonExecutable = process.platform === 'win32' ? 'python.exe' : 'python3';
  }

  outputChannel.appendLine(`Using Python executable: ${pythonExecutable} (venv: ${usingVenv})`);
  
  try {
    const stdout = await executeCommand(pythonExecutable, [pythonScriptPath, filePath], { showOutput: false });
    outputChannel.appendLine(`✅ Python script executed successfully`);
    return JSON.parse(stdout);
  } catch (error) {
    outputChannel.appendLine(`❌ Error executing Python DOM parser: ${error}`);
    
    // If virtual environment failed, try with system Python
    if (usingVenv) {
      outputChannel.appendLine('Retrying with system Python...');
      try {
        const systemPython = process.platform === 'win32' ? 'python.exe' : 'python3';
        const stdout = await executeCommand(systemPython, [pythonScriptPath, filePath], { showOutput: false });
        outputChannel.appendLine(`✅ System Python executed successfully`);
        return JSON.parse(stdout);
      } catch (systemError) {
        outputChannel.appendLine(`❌ System Python also failed: ${systemError}`);
        outputChannel.show(); // Show the output channel for debugging
        throw new Error(`Failed to execute Python DOM parser script. Error: ${systemError}`);
      }
    } else {
      outputChannel.show(); // Show the output channel for debugging
      throw new Error(`Failed to execute Python DOM parser script. Error: ${error}`);
    }
  }
}

/**
 * Parses HTML file and extracts DOM structure using Python parser
 * @param filePath Path to the HTML file
 * @returns DOM analysis result
 */
export async function parseHTMLFile(filePath: string): Promise<DOMAnalysisResult> {
  try {
    // Read the HTML file first for content
    const htmlContent = await fs.promises.readFile(filePath, 'utf-8');
    const fileName = path.basename(filePath);
    
    // Use Python script for DOM parsing
    const pythonResult = await callPythonDOMParser(filePath);
    
    if (pythonResult.error) {
      throw new Error(pythonResult.error);
    }
    
    const result: DOMAnalysisResult = {
      fileName,
      filePath,
      totalElements: pythonResult.totalElements,
      maxDepth: pythonResult.maxDepth,
      htmlContent: htmlContent,
      domTree: pythonResult.domTree,
      elementCounts: pythonResult.elementCounts,
      timestamp: new Date().toISOString()
    };
    
    return result;
  } catch (error) {
    throw new Error(`Failed to parse HTML file: ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Prepares HTML content for template injection
 * @param domAnalysis DOM analysis result
 * @returns Clean HTML ready for babia-html attribute (single line, no escaping)
 */
export function prepareHTMLForTemplate(domAnalysis: DOMAnalysisResult): string {
  let htmlContent = domAnalysis.htmlContent;
  
  // Extract only the body content, not the full HTML document
  const bodyMatch = htmlContent.match(/<body[^>]*>([\s\S]*?)<\/body>/i);
  if (bodyMatch) {
    htmlContent = bodyMatch[1].trim();
  } else {
    // If no body tag, try to extract content between <html> tags
    const htmlMatch = htmlContent.match(/<html[^>]*>([\s\S]*?)<\/html>/i);
    if (htmlMatch) {
      // Remove head section and keep only what would be in body
      let content = htmlMatch[1];
      const headMatch = content.match(/<head[^>]*>[\s\S]*?<\/head>/i);
      if (headMatch) {
        content = content.replace(headMatch[0], '').trim();
      }
      htmlContent = content;
    }
  }
  
  // Clean up the HTML content - CRITICAL: Convert to single line and remove problematic content
  let cleanHTML = htmlContent
    .replace(/<!--[\s\S]*?-->/g, '')           // Remove comments
    .replace(/<script[\s\S]*?<\/script>/gi, '') // Remove script tags
    .replace(/<style[\s\S]*?<\/style>/gi, '')   // Remove style tags
    .replace(/\r\n/g, ' ')                      // Replace Windows line endings with spaces
    .replace(/\n/g, ' ')                        // Replace Unix line endings with spaces  
    .replace(/\r/g, ' ')                        // Replace Mac line endings with spaces
    .replace(/\t/g, ' ')                        // Replace tabs with spaces
    .replace(/\s+/g, ' ')                       // Normalize multiple spaces to single space
    .trim();                                    // Remove leading/trailing whitespace
  
  // If the HTML is too long, truncate it but keep it valid
  if (cleanHTML.length > 3000) {
    // Find a good place to cut (try to end at a closing tag)
    const cutPoint = cleanHTML.lastIndexOf('>', 3000);
    if (cutPoint > 2500) {
      cleanHTML = cleanHTML.substring(0, cutPoint + 1);
    } else {
      cleanHTML = cleanHTML.substring(0, 3000);
    }
  }
  
  // Ensure we have some basic content if extraction failed
  if (!cleanHTML || cleanHTML.length < 10) {
    cleanHTML = `<div><h1>Sample Content</h1><p>HTML content extracted from ${domAnalysis.fileName}</p></div>`;
  }
  
  // CRITICAL: Return as single line without any escaping
  // The babia-html component expects raw HTML, not escaped HTML
  return cleanHTML;
}

File: src/new_code_analysis/python/html_dom_parser.py
#!/usr/bin/env python3
"""
HTML DOM Parser

This script parses HTML files and extracts DOM structure information.
It's designed to be called from the TypeScript code in the CodeXR extension.

Usage: python html_dom_parser.py <file_path>
"""

import sys
import json
import os
from html.parser import HTMLParser
from typing import Dict, List, Any, Optional


class DOMElement:
    """Represents a DOM element with its properties"""
    
    def __init__(self, tag_name: str, attributes: Dict[str, str], depth: int = 0):
        self.tag_name = tag_name.lower()
        self.attributes = attributes
        self.children: List['DOMElement'] = []
        self.text_content: str = ""
        self.depth = depth
        self.id = attributes.get('id')
        self.classes = attributes.get('class', '').split() if attributes.get('class') else []
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            'tagName': self.tag_name,
            'attributes': self.attributes,
            'children': [child.to_dict() for child in self.children],
            'textContent': self.text_content,
            'depth': self.depth,
            'id': self.id,
            'classes': self.classes
        }


class HTMLDOMParser(HTMLParser):
    """Custom HTML parser for extracting DOM structure"""
    
    def __init__(self):
        super().__init__()
        self.root: Optional[DOMElement] = None
        self.element_stack: List[DOMElement] = []
        self.current_depth = 0
        self.element_counts: Dict[str, int] = {}
        self.total_elements = 0
        self.max_depth = 0
        self.current_text = ""
    
    def handle_starttag(self, tag: str, attrs: List[tuple]):
        """Handle opening tags"""
        # Convert attributes to dictionary
        attributes = dict(attrs)
        
        # Create new element
        element = DOMElement(tag, attributes, self.current_depth)
        
        # Update statistics
        self.total_elements += 1
        self.max_depth = max(self.max_depth, self.current_depth)
        self.element_counts[tag.lower()] = self.element_counts.get(tag.lower(), 0) + 1
        
        # Add to parent or set as root
        if self.element_stack:
            self.element_stack[-1].children.append(element)
        else:
            self.root = element
        
        # Push to stack and increase depth
        self.element_stack.append(element)
        self.current_depth += 1
        
        # Clear any accumulated text
        self.current_text = ""
    
    def handle_endtag(self, tag: str):
        """Handle closing tags"""
        if self.element_stack and self.element_stack[-1].tag_name == tag.lower():
            # Add accumulated text content
            if self.current_text.strip():
                self.element_stack[-1].text_content = self.current_text.strip()
            
            # Pop from stack and decrease depth
            self.element_stack.pop()
            self.current_depth -= 1
            self.current_text = ""
    
    def handle_data(self, data: str):
        """Handle text content"""
        self.current_text += data
    
    def get_analysis_result(self, file_path: str) -> Dict[str, Any]:
        """Get the complete analysis result"""
        return {
            'fileName': os.path.basename(file_path),
            'filePath': file_path,
            'totalElements': self.total_elements,
            'maxDepth': self.max_depth,
            'domTree': self.root.to_dict() if self.root else None,
            'elementCounts': self.element_counts,
            'timestamp': '',  # Will be set by TypeScript
        }


def analyze_html_file(file_path: str) -> Dict[str, Any]:
    """Analyze an HTML file and return DOM structure"""
    try:
        # Check if file exists
        if not os.path.exists(file_path):
            return {"error": f"File not found: {file_path}"}
        
        # Read the HTML file
        with open(file_path, 'r', encoding='utf-8', errors='ignore') as file:
            html_content = file.read()
        
        # Parse the HTML
        parser = HTMLDOMParser()
        parser.feed(html_content)
        
        # Get analysis result
        result = parser.get_analysis_result(file_path)
        result['htmlContent'] = html_content
        
        return result
        
    except Exception as e:
        return {"error": f"Error parsing HTML file: {str(e)}"}


def prepare_html_for_template(html_content: str, file_name: str) -> str:
    """Prepare HTML content for babia-html template injection"""
    try:
        # Extract body content
        body_start = html_content.lower().find('<body')
        body_end = html_content.lower().find('</body>')
        
        if body_start != -1 and body_end != -1:
            # Find the end of the opening body tag
            body_tag_end = html_content.find('>', body_start)
            if body_tag_end != -1:
                html_content = html_content[body_tag_end + 1:body_end].strip()
        else:
            # Try to extract content between <html> tags and remove head
            html_start = html_content.lower().find('<html')
            html_end = html_content.lower().find('</html>')
            
            if html_start != -1 and html_end != -1:
                html_tag_end = html_content.find('>', html_start)
                if html_tag_end != -1:
                    content = html_content[html_tag_end + 1:html_end]
                    
                    # Remove head section
                    head_start = content.lower().find('<head')
                    head_end = content.lower().find('</head>')
                    
                    if head_start != -1 and head_end != -1:
                        head_end_tag = content.find('>', head_end)
                        if head_end_tag != -1:
                            content = content[:head_start] + content[head_end_tag + 1:]
                    
                    html_content = content.strip()
        
        # Clean up the HTML content
        import re
        
        # Remove comments, scripts, and styles
        html_content = re.sub(r'<!--[\s\S]*?-->', '', html_content)
        html_content = re.sub(r'<script[\s\S]*?</script>', '', html_content, flags=re.IGNORECASE)
        html_content = re.sub(r'<style[\s\S]*?</style>', '', html_content, flags=re.IGNORECASE)
        
        # Normalize whitespace
        html_content = re.sub(r'\r\n|\n|\r|\t', ' ', html_content)
        html_content = re.sub(r'\s+', ' ', html_content)
        html_content = html_content.strip()
        
        # Truncate if too long
        if len(html_content) > 3000:
            # Try to find a good cutting point
            cut_point = html_content.rfind('>', 0, 3000)
            if cut_point > 2500:
                html_content = html_content[:cut_point + 1]
            else:
                html_content = html_content[:3000]
        
        # Ensure we have some content
        if not html_content or len(html_content) < 10:
            html_content = f'<div><h1>Sample Content</h1><p>HTML content extracted from {file_name}</p></div>'
        
        return html_content
        
    except Exception as e:
        return f'<div><h1>Error</h1><p>Failed to process HTML: {str(e)}</p></div>'


def main():
    """Main entry point"""
    if len(sys.argv) < 2:
        error_msg = {"error": "No file path provided"}
        print(json.dumps(error_msg))
        sys.exit(1)

    file_path = sys.argv[1]
    
    # Check if this is a template preparation request
    if len(sys.argv) > 2 and sys.argv[2] == '--prepare-template':
        try:
            with open(file_path, 'r', encoding='utf-8', errors='ignore') as file:
                html_content = file.read()
            
            prepared_html = prepare_html_for_template(html_content, os.path.basename(file_path))
            result = {"preparedHTML": prepared_html}
            print(json.dumps(result))
        except Exception as e:
            error_msg = {"error": f"Failed to prepare HTML template: {str(e)}"}
            print(json.dumps(error_msg))
        return
    
    # Regular DOM analysis
    result = analyze_html_file(file_path)
    
    # Output as JSON
    print(json.dumps(result))


if __name__ == "__main__":
    main()
